modules/porous_flow/src/materials/PorousFlowFluidStateSingleComponent.C

```diff
@@ -126,261 +82,215 @@ PorousFlowFluidStateSingleComponent::PorousFlowFluidStateSingleComponent(
-        (*_dsaturation_dvar)[_qp][_gas_phase_number][_hvar];
-    (*_dgrads_qp_dgradv)[_qp][_aqueous_phase_number][_hvar] =
-        -(*_dgrads_qp_dgradv)[_qp][_gas_phase_number][_hvar];
-
-    const Real d2s_dp2 = (fsp_dp[_gas_phase_number].saturation.derivatives()[_pidx] -
-                          _fsp[_gas_phase_number].saturation.derivatives()[_pidx]) /
-                         dp;
-
-    const Real d2s_dh2 = (fsp_dh[_gas_phase_number].saturation.derivatives()[_hidx] -
-                          _fsp[_gas_phase_number].saturation.derivatives()[_hidx]) /
-                         dh;
-
-    const Real d2s_dph = (fsp_dp[_gas_phase_number].saturation.derivatives()[_hidx] -
-                          _fsp[_gas_phase_number].saturation.derivatives()[_hidx]) /
-                             (2.0 * dp) +
-                         (fsp_dh[_gas_phase_number].saturation.derivatives()[_pidx] -
-                          _fsp[_gas_phase_number].saturation.derivatives()[_pidx]) /
-                             (2.0 * dh);
-
-    (*_dgrads_qp_dv)[_qp][_gas_phase_number][_pvar] =
-        d2s_dp2 * _liquid_gradp_qp[_qp] + d2s_dph * _gradh_qp[_qp];
-    (*_dgrads_qp_dv)[_qp][_aqueous_phase_number][_pvar] =
-        -(*_dgrads_qp_dv)[_qp][_gas_phase_number][_pvar];
-
-    (*_dgrads_qp_dv)[_qp][_gas_phase_number][_hvar] =
-        d2s_dh2 * _gradh_qp[_qp] + d2s_dph * _liquid_gradp_qp[_qp];
-    (*_dgrads_qp_dv)[_qp][_aqueous_phase_number][_hvar] =
-        -(*_dgrads_qp_dv)[_qp][_gas_phase_number][_hvar];
-
-    // Gradient of porepressure and derivatives
-    // Note: need first and second derivativea of capillary pressure
-    const Real dpc = _pc.dCapillaryPressure(_fsp[_aqueous_phase_number].saturation.value());
-    const Real d2pc = _pc.d2CapillaryPressure(_fsp[_aqueous_phase_number].saturation.value());
-
-    (*_gradp_qp)[_qp][_aqueous_phase_number] = _liquid_gradp_qp[_qp];
-    (*_gradp_qp)[_qp][_gas_phase_number] =
-        _liquid_gradp_qp[_qp] + dpc * (*_grads_qp)[_qp][_aqueous_phase_number];
-
-    for (unsigned int ph = 0; ph < _num_phases; ++ph)
-      (*_dgradp_qp_dgradv)[_qp][ph][_pvar] = 1.0;
-
-    (*_dgradp_qp_dgradv)[_qp][_gas_phase_number][_pvar] +=
-        dpc * (*_dgrads_qp_dgradv)[_qp][_aqueous_phase_number][_pvar];
-    (*_dgradp_qp_dgradv)[_qp][_gas_phase_number][_hvar] =
-        dpc * (*_dgrads_qp_dgradv)[_qp][_aqueous_phase_number][_hvar];
-
-    (*_dgradp_qp_dv)[_qp][_gas_phase_number][_pvar] =
-        d2pc * (*_grads_qp)[_qp][_aqueous_phase_number] *
-            (*_dsaturation_dvar)[_qp][_aqueous_phase_number][_pvar] +
-        dpc * (*_dgrads_qp_dv)[_qp][_aqueous_phase_number][_pvar];
-
-    (*_dgradp_qp_dv)[_qp][_gas_phase_number][_hvar] =
-        d2pc * (*_grads_qp)[_qp][_aqueous_phase_number] *
-            (*_dsaturation_dvar)[_qp][_aqueous_phase_number][_hvar] +
-        dpc * (*_dgrads_qp_dv)[_qp][_aqueous_phase_number][_hvar];
-  }
+    if constexpr (!is_ad)
```

Comment:
could be an idea to propagate that to the other if (the use of constexpr)

Reply:
why do you not need second derivatives with AD? AD doesnt do that for you does it?